README.md
<p align="center">
<a href="#"><img src="http://i.imgur.com/XkEopOD.png"/></a>

<a href="#install"><img src="https://img.shields.io/badge/platform-android-brightgreen.svg?style=flat-square"/></a>
<a href="https://nodejs.org/en/"><img src="https://img.shields.io/badge/server%20platform-nodejs-brightgreen.svg?style=flat-square"/></a>
<a href="https://play.google.com/store/apps/details?id=air.MultiplayerMagic"><img src="https://img.shields.io/badge/installs-50-brightgreen.svg?style=flat-square"/></a>
<a href="https://github.com/AndrewK9/Multiplayer-Magic/issues"><img src="https://img.shields.io/badge/issues-0%20open-brightgreen.svg?style=flat-square"/></a>
<a href="https://github.com/AndrewK9/Multiplayer-Magic/blob/master/LICENSE"><img src="https://img.shields.io/badge/license-MIT-lightgray.svg?style=flat-square"/></a>
</p>

## Table of Contents
1. [Introduction](#introduction)
1. [Install](#install)
1. [Usage](#usage)
1. [Team](#team)
1. [Contribute](#contribute)
1. [License](#license)

## Introduction
The goal of Multiplayer Magic was to create an android application where each user can manage their own life and infect totals while being able to check the other players stats. I created an app where the users can connect to a server, input their names and join or host a match. People can play with their friends and focus on having fun instead of forcing one player to manage everyones stats.

## Install
This project uses [Node.js](https://nodejs.org/en/) and no non-standard dependencies.

You can download the client for Android [here](https://play.google.com/store/apps/details?id=air.MultiMTG&hl=en).

## Usage
### Server
To launch the server just run the server.js file with nodejs. You could also connect to the US based server if you can't host your own.

### Players
Players can join a match by typing in the custom room code.

## Team
#### Creator

Kyle Andrews |
|-----|
| [Email](andrewskyle28@gmail.com) |
| @andrewk9 |

## Contribute
Feel free to [open an issue](https://github.com/AndrewK9/Multiplayer-Magic/issues) or clone this repo and start working.

## License
[MIT](https://github.com/AndrewK9/Multiplayer-Magic/blob/master/LICENSE) (c) Kyle Andrews


CHANGELOG.md
# Change Log
All notable chagnes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

---

## [2.1.1] - 2017-02-18
### Added
- New Icon.png images for desktop version.

### Changed
- Updated the APK and SWF files.
- Updated the Google Play store link for the README file.

## [2.1.0] - 2017-02-17 [RELEASE]

## [2.0.4] - 2017-02-17
### Changed
- Server only allows 1 restart packet, ignores the rest.

## [2.0.3] - 2017-02-17
### Changed
- Updated PROTOCOL.md

## [2.0.2] - 2017-02-17
### Added
- On invalid match message we add text to show that players can host a match.

### Changed
- Changed text to static for 'Input Username or Initials' and 'Match Code'.

## [2.0.1] - 2017-02-17
### Added
- When clients join mid match we send a _MMIR_ Mid Match Info Request.
- The server will send them _MMJR_ Mid Match Join Responce.
- Added a releaseTheClientsMidMatch function that acts like releaseTheClients where it sends player info however this one only sends the players info to the single client who requested it.

### Changed
- The server no longer sends the Join Responce packet by default, it now sends 3 possible types.
	1: Join Responce is the normal responce for when players join an open match.
	2: Match Error will return the error message.
	3: Join Mid Match tells the spectator that they can join mid match.

## [2.0.0] - 2017-02-16 [RELEASE]

## [1.2.5] - 2017-02-16
### Added
- In GSMatch there is a _Loading Players_ message until we unlock input.
- Updated name to Multiplayer Magic.
- Updated US-East IP info.

### Changed
- Reset the Matche classes max players back to 8.
- Updated the PROTOCOL.md file.
- Generated Protocol.pdf.
- Updated README.md file.

## [1.2.4] - 2017-02-16
### Added
- The Match class now sends an _UNLK_ Unlock Input packet to all the players saying that they can add the event listeners to their buttons.
- The client now handles _UNLK_ packets.

### Changed
- New text size for the match code on GSLobby, was 80 and is now 75.
- Reduced the font size for the clients chatroom message input.
- Message input is now the length of the screen.
- The chat button text has ALT symbol arrows that update with the functionality of the button.
- Adjusted the position of the chat button.
- The app no longer forces full screen on mobile.
- The chat toggle button now says _- TOGGLE CHAT -_.
- We now scale content with the stage.

## [1.2.3] - 2017-02-15
### Added
- Added scrollV to the chat text box to auto scroll when new messages arrive.

### Changed
- Stoped using HTML text due to a possible issue with the android verson. We now user the basic .text when addings messages to the chat.

## [1.2.2] - 2017-02-15
### Added
- Fixed issue by setting client.match equal to the match in the checkForMatch function for spectators.
- Added a chatMessage function to the Match class, it loops through all the spectators and broadcast the chat messages.
- Connection.as parses message packets.
- Set max characters on the chat messages to 120.
- Connection.as passes the chat messages to the GSMatch screen so we can display the messages.

## [1.2.1] - 2017-02-15
### Added
- Designed the spectator chatroom.
- The chat button will toggle the chatroom into position.

## [1.2.0] - 2017-02-15
### Added
- Merged feat/BattleScreen with the development for version 1.2.0

## [1.1.18] - 2017-02-15
### Added
- broadcastRageQuit sends an update packet to everyone, the player symbol won't go away but it will say they are dead.
- After handling input the Match class checks for death. If someone dies we run the checkForWin function.
- checkForWin loops through all the players and counts the deaths, if 0 to 1 player is alive we declare a winner.
- The broadcastGameOver function loops through everyone connected to the match and sends a _GMOV_ packet.
- The Game Over packet contains the name of the winner.
- Connection.as handels _GMOV_ packets by sending the winner to Game.as and then to GSMatch.
- GSMatch clears the screen with the gameOver function.
- We spawn a Winner symbol and pass the winners name into it.
- Winner has two buttons, players can quit the app or restart the match.
- gameOver hides the buttons if they're a spectator.

## [1.1.17] - 2017-02-15
### Chagned
- The readPacketUpdate, readPacketPrivateUpdate, and readPacketStartUpdate all call the tryReadingPacket function when they are done. Mobile devices had lag and a major buffer backlog and becuase the tryReadingPacket function doesn't endlessly loop the player would get behind on updates.
- tryReadingPacket now checks the length of the buffer, if it's less than 4 we return out of the function.

## [1.1.16] - 2017-02-15
### Added
- Connection.as handles update packets and passes the parsed info to Game.as who trys to send it to the update function in GSMatch.
- When loading players we now assign the txtHealth and txtInfect text boxes to the correct values.
- GSMatch now has an update function that loops through all the players and finds the matching ID. We when call that players update function and pass in the new health and infect numbers.
- When broadcasting update we send a private update to the player who sent the input.

server.js
const MTGP = {
	NAME_SHORT: 1,
	NAME_LONG: 2,
	NAME_INVALID: 3,
	GOOD: 0,
	MATCH_FULL: 1,
	MATCH_INVALID: 2,
	MATCH_FULL_STARTED: 5,
	buildNameError: (err) =>{
		const packet = Buffer.alloc(5);
		packet.write("NERR");
		packet.writeUInt8(err, 4);
		return packet;
	},
	buildMatchError: (err) =>{
		const packet = Buffer.alloc(5);
		packet.write("MERR");
		packet.writeUInt8(err, 4);
		return packet;
	},
	buildJoinResponce: (responce, matchCode, numOfPlayers) =>{
		const packet = Buffer.alloc(12);
		packet.write("MJRS");
		packet.writeUInt8(responce, 4);
		packet.writeUInt8(numOfPlayers, 5);
		packet.write(matchCode, 6);
		return packet;
	},
	buildHostResponce: (matchCode) =>{
		const packet = Buffer.alloc(10);
		packet.write("MHRS");
		packet.write(matchCode.toUpperCase(), 4);
		return packet;
	},
	buildStartPacket: (isPlayer) =>{
		const packet = Buffer.alloc(5);
		packet.write("PSTG");
		packet.writeUInt8(isPlayer ? 1 : 0, 4);
		return packet;
	},
	buildLobbyUpdate: (numOfPlayers) =>{
		const packet = Buffer.alloc(5);
		packet.write("LUPD");
		packet.writeUInt8(numOfPlayers, 4);
		return packet;
	},
	buildUpdate: (client)=>{
		const packet = Buffer.alloc(7);
		packet.write("UPDT");
		packet.writeUInt8(client.playerid, 4);
		packet.writeUInt8(client.health, 5);
		packet.writeUInt8(client.infect, 6);
		return packet;
	},
	buildPrivateUpdate: (client)=>{
		const packet = Buffer.alloc(6);
		packet.write("PUDT");
		packet.writeUInt8(client.health, 4);
		packet.writeUInt8(client.infect, 5);
		return packet;
	},
	buildStartUpdatePacket: (player, match) => {
		const packet = Buffer.alloc(16);
		packet.write("GSUD");
		packet.writeUInt8(player.playerid, 4);
		packet.writeUInt8(player.health, 5);
		packet.writeUInt8(player.infect, 6);
		packet.writeUInt8(match.maxInfect, 7)
		packet.write(player.username, 8);
		return packet;
	},
	buildGameOver: (winner)=>{
		const packet = Buffer.alloc(12);
		packet.write("GMOV");
		packet.write(winner, 4);
		return packet;
	},
	buildChat: (message, username)=>{
		const packet = Buffer.alloc(133);
		packet.write("BMSG");
		packet.write(username, 4);
		packet.write(message, 12);
		return packet;
	},
	buildUnlockPacket: ()=>{
		const packet = Buffer.alloc(4);
		packet.write("UNLK");
		return packet;
	},
	buildJoinMidMatch: ()=>{
		const packet = Buffer.alloc(4);
		packet.write("JMMR");
		return packet;
	},
};

class Server {
	constructor(){
		this.port = 2231;
		this.clients = [];
		this.matches = [];
		this.id = 0;

		this.sock = require("net").createServer((sock) => {
			this.clients.push(new Client(this, sock, this.id));
			this.id++;
		});
		this.sock.listen(this.port, () => {
			console.log("++[STARTRING SERVER]++");
			console.log("[SERVER] The server is running on port " + this.port);
		});
	}
	handleDisconnect(client){
		console.log("[SERVER] "+client.username+" disconnected (IP:" + client.sock.remoteAddress + ")");
		//Check to see if they were in a match
		if(client.matchCode != ""){
			//Remove them from the match
			this.removeFromMatch(client);
		}
		this.clients.splice(this.clients.indexOf(client), 1);
	}
	removeFromMatch(client){
		//We will loop though all the matches looking for one that matches the clients match code
		this.matches.map((match) => {
			if(match.code.toUpperCase() == client.matchCode.toUpperCase()){
				if(client.isPlayer){
					//The player is a client, we should update the lobby
					match.players.splice(match.players.indexOf(client), 1);
					console.log("["+match.code.toUpperCase()+"] Has " + match.players.length + "/" + match.maxPlayers + " players");
					if(!match.hasStarted) this.broadcastNewGSLobbyPlayer(match.code, match.players.length);
					else match.broadcastRageQuit(client);
					if(match.players.length <= 0) {
						//The match is now empty, let's remove it
						console.log("["+match.code.toUpperCase()+"] Is empty and has been removed");
						this.matches.splice(this.matches.indexOf(match), 1);
						//console.log(this.matches.length);
					}
				}else{
					//The client is a spectator, we will siently remove them
					match.spectators.splice(match.spectators.indexOf(client), 1);
					console.log("["+match.code.toUpperCase()+"] " + client.username + " has stopped spectating the match");
				}
			}
		});
	}
	isNameOkay(username){
		if(username.length < 2) return MTGP.NAME_SHORT;
		if(username.length > 8) return MTGP.NAME_LONG;
		if(!username.match(/^[a-zA-Z0-9\s\.\-\_]+$/)) return MTGP.NAME_INVALID;

		return MTGP.GOOD;
	}
	checkForMatch(matchCode, client){
		//console.log(">"+matchCode+"<");
		//console.log(this.matches.length);
		let foundMatch = false;
		let matchStarted = false;
		let isPlayer = false;
		if(!matchCode.match(/^[a-zA-Z]+$/)) return MTGP.MATCH_INVALID;
		this.matches.map((match) => {
			if(matchCode.toUpperCase() == match.code.toUpperCase()) {
				if(match.players.length < match.maxPlayers && !match.hasStarted) {
					match.players.push(client);
					client.isPlayer = true;
					client.match = match;
					console.log("["+match.code.toUpperCase()+"] " + client.username + " joined the match");
					console.log("["+match.code.toUpperCase()+"] Has " + match.players.length + "/" + match.maxPlayers + " players");
					foundMatch = true;
					isPlayer = true;
					//console.log(match.players.length);
					if(match.players.length > 1) this.broadcastNewGSLobbyPlayer(matchCode, match.players.length);
				}else{
					//The new player is a spectator
					match.spectators.push(client);
					client.match = match;
					console.log("["+match.code.toUpperCase()+"] " + client.username + " is spectating the match");
					foundMatch = true;
					matchStarted = match.hasStarted;
				}
			}
		});

		if(!foundMatch) { return MTGP.MATCH_INVALID; } //We couldn't find a match
		else if(!isPlayer && !matchStarted) { return MTGP.MATCH_FULL; } //The match was full so the client is a spectator
		else if(!isPlayer && matchStarted) { return MTGP.MATCH_FULL_STARTED; }
		else { return MTGP.GOOD; } //We found the match and there was space for another player
	}
	broadcastNewGSLobbyPlayer(code, numOfPlayers){
		this.clients.map((client)=>{
			if(client.matchCode.toUpperCase() == code.toUpperCase()){
				//This client is in the match, tell them it started
				client.sock.write(MTGP.buildLobbyUpdate(numOfPlayers));
			}
		});
	}
	createMatch(){
		const matchCode = this.generateMatchCode();
		this.matches.push(new Match(matchCode));
		return matchCode;
	}
	generateMatchCode(){
		const letter1 = this.getRandomInt(1,26);
		const letter2 = this.getRandomInt(1,26);
		const letter3 = this.getRandomInt(1,26);
		const letter4 = this.getRandomInt(1,26);
		const letter5 = this.getRandomInt(1,26);
		const letter6 = this.getRandomInt(1,26);

		var matchCode = "";
		matchCode = this.getLetterFromInt(letter1)+
		this.getLetterFromInt(letter2)+
		this.getLetterFromInt(letter3)+
		this.getLetterFromInt(letter4)+
		this.getLetterFromInt(letter5)+
		this.getLetterFromInt(letter6);

		console.log("[SERVER] New host generated match code: " + matchCode.toUpperCase());

		return matchCode;
	}
	getRandomInt(min, max){
		return Math.floor(Math.random() * (max - min + 1)) + min;
	}
	getLetterFromInt(number){
		switch(number){
        	case 1:
        	    return "a";
        	case 2:
        	    return "b";
        	case 3:
        	    return "c";
        	case 4:
        	    return "d";
        	case 5:
        	    return "e";
        	case 6:
        	    return "f";
        	case 7:
        	    return "g";
        	case 8:
        	    return "h";
        	case 9:
        	    return "i";
        	case 10:
        	    return "j";
        	case 11:
        	    return "k";
        	case 12:
        	    return "l";
        	case 13:
        	    return "m";
        	case 14:
        	    return "n";
        	case 15:
        	    return "o";
        	case 16:
        	    return "p";
        	case 17:
        	    return "q";
        	case 18:
        	    return "r";
        	case 19:
        	    return "s";
        	case 20:
        	    return "t";
        	case 21:
        	    return "u";
        	case 22:
        	    return "v";
        	case 23:
        	    return "w";
        	case 24:
        	    return "x";
        	case 25:
        	    return "y";
        	case 26:
        	    return "z";
    	}
	}
	attemptMatchStart(code, client){
		this.matches.map((match)=>{
			if(code.toUpperCase() == match.code.toUpperCase()){
				//We found the match, we need to check if there are enough players
				if(match.players.length >= 2){
					//The match can start
					match.hasStarted = true;
					console.log("["+match.code.toUpperCase()+"] " + client.username + " has started the match");
					//TODO: Send clients a start packet
					match.broadcastStartMatch();
					match.calculateMatch();
					//We have to set a timeout because not all clients are loading into the match before we start sending them the other clients.
					setTimeout(()=>{
						match.releaseTheClients();
					}, 3000);
				}else{
					//The match can't start
					console.log("["+match.code.toUpperCase()+"] Not enough players to start a match");
				}
			}
		});
	}
	getLobbyNumber(code){
		let numOfPlayers = 0;
		this.matches.map((match)=>{
			if(match.code.toUpperCase() == code.toUpperCase()){
				//We found the match, return the number
				numOfPlayers =  match.players.length;
				return;
			}
		});

		return numOfPlayers;
	}
}

class Client {
	constructor(server, sock, id){
		
		this.playerid = id;
		this.server = server;
		this.sock = sock;
		this.buffer = Buffer.alloc(0);
		this.username = "";
		this.matchCode = "";
		this.isPlayer = false;
		this.match = null;
		this.health = 0;
		this.infect = 0;
		this.isDead = true;

		this.sock.on('error', (msg) => {});
		this.sock.on('close', () => { this.server.handleDisconnect(this); });
		this.sock.on('data', (data) => this.onData(data));

		console.log("[SERVER] A new client connected (IP:" + sock.remoteAddress + ", ID: " + this.playerid + ")");
	}
	onData(data){
		console.log("[SERVER] Handling incoming data");
		this.buffer = Buffer.concat([this.buffer, Buffer.from(data)]);

		while(this.buffer.length > 0){
			if(this.tryReadingPacket()) { break; }
			this.destroyStreamData();
		}
	}
	destroyStreamData(){
		this.buffer = this.buffer.slice(1, this.buffer.length);
		console.log("[SERVER] Destroying stream data");
	}
	tryReadingPacket(){
		console.log("[SERVER] Trying to read the packet");
		switch(this.getNextPacketType()){
			case null:
				break;
			case "JOIN":
				this.readPacketJoin();
				break;
			case "HOST":
				this.readPacketHost();
				break;
			case "UMSR":
				this.readPacketStart();
				break;
			case "UIUP":
				this.readPacketInput();
				break;
			case "REST":
				this.readPacketRestart();
				break;
			case "UMSG":
				this.readPacketUserMessage();
				break;
			case "MMIR":
				this.readPacketInfoRequest();
				break;
			default:
				return false;
				break;
		}
		return true;
	}
	getNextPacketType(){
		console.log("[SERVER] Grabbing packet type");
		if(this.buffer.length < 4) return null;
		console.log("[SERVER] Packet type: " + this.buffer.slice(0, 4).toString());
		return this.buffer.slice(0, 4).toString();
	}
	splitBufferAt(n){
		this.buffer = this.buffer.slice(n, this.buffer.length);
	}
	readPacketJoin(){
		console.log("[SERVER] Parsing the join packet");
		//console.log(this.buffer.length);
		if(this.buffer.length < 11) return;
		const matchCode = this.buffer.slice(4, 10).toString();
		const usernameLength = this.buffer.readUInt8(10);
		const packetLength = 11 + usernameLength;
		//console.log(this.buffer.length + ":" + packetLength);
		if(this.buffer.length < packetLength) return;
		const username = this.buffer.slice(11, 11 + usernameLength).toString();
		this.splitBufferAt(packetLength);
		//console.log(">" + matchCode + "<>" + username + "<");

		//Handle the username
		let errorCode = this.server.isNameOkay(username);
		if(errorCode === 0) { this.username = username; }
		else { 
			this.sock.write(MTGP.buildNameError(errorCode));
			console.log("[ERROR] " + username + " is an invalid username");
			return;
		}

		//Check to see if the match exist
		let matchResponce = this.server.checkForMatch(matchCode, this);
		//console.log(matchResponce);
		if(matchResponce === 0 || matchResponce === 1) {
			this.matchCode = matchCode;
			this.sock.write(MTGP.buildJoinResponce(matchResponce, this.matchCode, this.server.getLobbyNumber(this.matchCode)));
			//console.log(this.matchCode.toUpperCase());
		}
		else if(matchResponce === 2) { 
			this.sock.write(MTGP.buildMatchError(matchResponce));
			console.log("[ERROR] " + matchCode.toUpperCase() + " is invalid");
			return;
		}
		else if(matchResponce === 5){
			this.matchCode = matchCode;
			this.sock.write(MTGP.buildJoinMidMatch());
		}
	}
	readPacketHost(){
		if(this.buffer.length < 5) return;
		const usernameLength = this.buffer.readUInt8(4);
		const packetLength = 5 + usernameLength;
		if(this.buffer.length < packetLength) return;
		const username = this.buffer.slice(5, 5 + usernameLength).toString();
		this.splitBufferAt(packetLength);
		console.log(">" + username + "<");

		//Handle the username
		let errorCode = this.server.isNameOkay(username);
		if(errorCode === 0) { this.username = username; }
		else { 
			this.sock.write(MTGP.buildNameError(errorCode));
			console.log("[ERROR] " + username + " is invalid");
			return;
		}
		//If we made it this far it means the username was legit
		//Create match
		this.matchCode = this.server.createMatch();
		let matchResponce = this.server.checkForMatch(this.matchCode, this);
		//Inform the player that they are good to go
		this.sock.write(MTGP.buildHostResponce(this.matchCode));
	}
	readPacketStart(){
		this.splitBufferAt(4);
		//console.log("A player tried to start, let's see what happens.");
		//A player has pressed the start button, we need to try to start thier match
		this.server.attemptMatchStart(this.matchCode, this);
		
	}
	readPacketInput(){
		//console.log(this.buffer.length);
		if(this.buffer.length < 5) return;
		const inputType = this.buffer.readUInt8(4);
		this.splitBufferAt(5);
		//console.log("I am reading packet input: " + inputType);
		//Now that we have the input type, we can pass it to the server
		this.match.handlePlayerInput(inputType, this);
	}
	readPacketUserMessage(){
		if(this.buffer.length < 6) return;
		const messageLength = this.buffer.readUInt8(4);
		const message = this.buffer.slice(5, 5 + messageLength).toString();
		this.splitBufferAt(5 + messageLength);
		//console.log(">"+message+"<");
		this.match.chatMessage(message, this);
	}
	readPacketRestart(){
		if(this.buffer.length < 4) return;
		this.splitBufferAt(4);
		//console.log(this.matchCode);
		if(!this.match.hasStarted) this.match.handleRestart();
	}
	readPacketInfoRequest(){
		setTimeout(()=>{
			this.match.releaseTheClientsMidMatch(this);
		}, 3000);
	}
}

class Match{
	constructor(matchCode){
		this.code = matchCode;
		this.maxPlayers = 8;
		this.players = [];
		this.spectators = [];
		this.hasStarted = false;
		this.startingHealth = 0;
		this.maxInfect = 0;
	}
	handlePlayerInput(inputType, client){
		//TODO: Check for input type with a switch and handle it
		switch(inputType){
			case 1:
				this.minusHealth(client);
				break;
			case 2:
				this.addHealth(client);
				break;
			case 3:
				this.minusInfect(client);
				break;
			case 4:
				this.addInfect(client);
				break;
			default:
				console.log("["+this.code.toUpperCase()+"] Recieved an unknown input type");
				break;
		}
	}
	minusHealth(client){
		client.health--;
		if(client.health < 0) client.health = 0;
		console.log("["+this.code.toUpperCase()+"] " + client.username + " lost health " + client.health);
		this.broadcastUpdate(client);
		this.checkForDeath(client);
	}
	addHealth(client){
		client.health++;
		console.log("["+this.code.toUpperCase()+"] " + client.username + " gained health " + client.health);
		this.broadcastUpdate(client);
		this.checkForDeath(client);
	}
	minusInfect(client){
		client.infect--;
		if(client.infect < 0) client.infect = 0;
		console.log("["+this.code.toUpperCase()+"] " + client.username + " lost infect " + client.infect);
		this.broadcastUpdate(client);
		this.checkForDeath(client);
	}
	addInfect(client){
		client.infect++;
		console.log("["+this.code.toUpperCase()+"] " + client.username + " gained infect " + client.infect);
		this.broadcastUpdate(client);
		this.checkForDeath(client);
	}
	checkForDeath(client){
		if(client.health <= 0 || client.infect >= this.maxInfect) {
			client.isDead = true;
			this.checkForWin();
			console.log("["+this.code.toUpperCase()+"] " + client.username + " died");
		}
		else client.isDead = false;

		
	}
	checkForWin(){
		let numOfDead = 0;
		this.players.map((player)=>{
			if(player.isDead) numOfDead++;
		});

		if(numOfDead >= this.players.length - 1){
			//All or all but one players are dead
			let winner = "No one";
			this.players.map((player)=>{
				if(!player.isDead) {
					winner = player.username;
					console.log("["+this.code.toUpperCase()+"] " + player.username + " won the game");
				}
			});

			this.broadcastGameOver(winner);
		}
	}
	broadcastGameOver(winner){
		this.hasStarted = false;
		this.players.map((player)=>{
			player.sock.write(MTGP.buildGameOver(winner));
		});

		this.spectators.map((spec)=>{
			spec.sock.write(MTGP.buildGameOver(winner));
		});
	}
	broadcastUpdate(client){
		client.sock.write(MTGP.buildPrivateUpdate(client));

		this.players.map((player)=>{
			if(player != client){
				player.sock.write(MTGP.buildUpdate(client));
			}
		});

		this.spectators.map((spec)=>{
			spec.sock.write(MTGP.buildUpdate(client));
		});
	}
	calculateMatch(){
		this.startingHealth = this.players.length * 10;
		this.maxInfect = this.startingHealth/2;
		console.log("["+this.code.toUpperCase()+"] Generated a starting health of: " + this.startingHealth + " and max infect of: " + this.maxInfect);
		this.players.map((player)=>{
			player.health = this.startingHealth;
			player.infect = 0;
			player.isDead = false;
		});
	}
	releaseTheClients(){
		//console.log("Updating all players about each other");
		this.players.map((player1)=>{
			this.players.map((player2)=>{
				if(player1 != player2){
					player1.sock.write(MTGP.buildStartUpdatePacket(player2, this));
				}
			});
		});

		this.spectators.map((spec)=>{
			this.players.map((player)=>{
				spec.sock.write(MTGP.buildStartUpdatePacket(player, this));
			});
		});

		this.unlockInput();
	}
	releaseTheClientsMidMatch(client){
		this.players.map((player)=>{
			client.sock.write(MTGP.buildStartUpdatePacket(player, this));
		});
	}
	unlockInput(){
		//Delay the message so the clients can handle all the new players
		//Should replace with a listerner where we wait for all players to return a ready message
		setTimeout(()=>{
			this.players.map((player)=>{
				player.sock.write(MTGP.buildUnlockPacket());
			});
		}, 2000);
	}
	broadcastStartMatch(){
		//console.log("Finding clients in the match and telling them to start");

		this.players.map((player)=>{
			player.sock.write(MTGP.buildStartPacket(player.isPlayer));
		});

		this.spectators.map((spec)=>{
			spec.sock.write(MTGP.buildStartPacket(spec.isPlayer));
		});
	}
	broadcastRageQuit(client){
		client.isDead = true;
		client.health = 0;

		this.players.map((player)=>{
			if(player != client){
				player.sock.write(MTGP.buildUpdate(client));
			}
		});

		this.spectators.map((spec)=>{
			spec.sock.write(MTGP.buildUpdate(client));
		});

		this.checkForWin();
	}
	handleRestart(){
		this.hasStarted = true;
		console.log("["+this.code.toUpperCase()+"] Someone restarted the match");
		this.broadcastStartMatch();
		this.calculateMatch();
		//We have to set a timeout because not all clients are loading into the match before we start sending them the other clients.
		setTimeout(()=>{
			this.releaseTheClients();
		}, 3000);
	}
	chatMessage(message, client){
		console.log("["+this.code.toUpperCase()+"] " + client.username + ": " + message);
		this.spectators.map((spec)=>{
			spec.sock.write(MTGP.buildChat(message, client.username));
		});
	}
}

new Server();